[SPARK-17330] [SPARK UT] Clean up spark-warehouse in UT

tone-zhang opened the pull request:
## What changes were proposed in this pull request?

Check the database warehouse used in Spark UT, and remove the existing database file before run the UT (SPARK-8368).
## How was this patch tested?

Run Spark UT with the command for several times:
./build/sbt -Pyarn -Phadoop-2.6 -Phive -Phive-thriftserver "test-only _HiveSparkSubmitSuit_"
Without the patch, the test case can be passed only at the first time, and always failed from the second time. 
With the patch the test case always can be passed correctly.

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented:
How would this relate to SPARK-8368?

We already have a Utils method for deleting recursively already; please use that.

Is this the only path that leaves behind spark-warehouse?
If it's definitely cleaned up we should remove it from .gitignore

tone-zhang commented:
@srowen Thanks for the comments.
I wrote SPARK-8368 here just because the UT case name is "SPARK-8368: includes jars passed in through --jars". 
For the Utils method you mentioned, could you please share me more hints? Do you mean I just need to remove "spark-warehouse" from .gitignore? If it is, I will have a try, and update the PR.
Thanks a lot!

srowen commented:
OK, but the change isn't actually related to SPARK-8368, so I think you can remove that link in the title and the link that was created in the JIRA.

See `Utils.deleteRecursively`. You don't need to implement it yourself.

No, I'm saying that _if_ you find all instances of tests that don't clean up spark-warehouse (there may be more, and we should try to fix all of them), then there's also no reason to have to ignore it in git. But that's not the fix, just a cleanup.

tone-zhang commented:
@srowen Thanks a lot for your help.
I have updated the code and used the Utils.deleteRecursively, the Utils method is very helpful. Thanks!
For the time being, I found the UT case is the only one which is impacted. 
I will update the title in Jira for #17330, and remove the link in *8368*.
For the updated code, could you please have a review?
Thanks a lot!

srowen commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSparkSubmitSuite.scala:
```diff
@@ -109,6 +109,8 @@ class HiveSparkSubmitSuite
   }
 
   test("SPARK-8368: includes jars passed in through --jars") {
+    val warehousePath = System.getProperty("user.dir") + "/../../spark-warehouse"
```
This seems like a bandaid. Where is this directory created and not cleaned up in the first place? I see other tests that create it, so, it's probably best to fix the source?

tone-zhang replied:
@srowen Thanks a lot for your help!
Yes, it is better to clean up the temporary path at the source. I will have a check and update the PR.
Thanks!

tone-zhang commented:
In Spark UT, the temporary data file should be stored in spark/target/tmp not in spark/spark-warehouse. The Spark UT suite will clean up the path spark/target/tmp when finish the test. It is the existing mechanism. 
In some test cases, the property "spark.sql.warehouse.dir" of SQLConf is not set, as default, the spark/spark-warehouse will be used to store the temporary data file, it is in correct.

@srowen Thanks a lot for your comments. Could you please have a review?  Thanks!

srowen commented:
This is better. It looks like `DDLSuite` tests the default location which is `./spark-warehouse` in "Create Database using Default Warehouse Path". That's fine but it fails to clean up the directory. I think it needs to clean up there too. You can see that elsewhere it uses this `withTempDir { tempDir =>` idiom to correctly set this to a temp dir.

Although I agree with your changes, aren't there other places in the same file that need the same treatment?

Ultimately the goal is to see that after a successful test run, no `spark-warehouse` exists. If that's true then we can remove it from `.gitignore` and `.rat-excludes`. CC @liancheng re: https://github.com/apache/spark/commit/63db2bd283a430971d85f2a7b06dac77723c56fa

tone-zhang commented:
@srowen I checked DDLSuite.scala, the case "Create Database using Default Warehouse Path" is used to test the default "spark-warehouse" setting. In the doc, the default path should be "("user.dir")/spark-warehouse". I prefer to keep the path, just remove it after finish the case manually. 
For the "tempDir" in DDLSuit.scala, I confirm it is point to the temporary path. It is safe enough.
When check  the cases in DDLSuite.scala, I found the "spark-warehouse" is always created by the UT frame. At the final, the directory is empty and it will not negative impact on the UT. I tried to set the WAREHOUSE_PATH of SQLConf to redirect to the temporary path, but failed. May I delete the "spark-warehouse" in function afterEach()? Or could you please give me some suggestion? Thanks a lot!

srowen commented:
Yes, I think it wouldn't hurt to always try to delete the default spark-warehouse in this suite after each test. Try running all tests after you're done to see if anything leaves it behind; the goal is to get rid of the entry in gitignore.

Commit:
[SPARK-17330] Clean up the temporary data files used in Spark UT

When run Spark UT based on the latest version of master branch,
One UT case is failed becaues the temporary data files have not been
cleaned up.
The UT name is "SPARK-8368: includes jars passed in through --jars".
In Spark UT, the temorary data file should be stored in spark/target/tmp
not in spark/spark-warehouse.

tone-zhang commented:
Update the PR to clean up the "spark-warehouse" in DDLSuite.scala.
@srowen I have run the whole Spark UT with the PR, I cannot see "spark/spark-warehoues" after the test. It seems reaches the goal.
Could you please have a review? Thanks a lot!

srowen commented:
Jenkins test this please

SparkQA commented:
**[Test build #65205 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65205/consoleFull)** for PR 14894 at commit [`cd8e9a4`](https://github.com/apache/spark/commit/cd8e9a4ddf704f2d01df870cc898af53e62a9d2f).

SparkQA commented:
**[Test build #65205 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65205/consoleFull)** for PR 14894 at commit [`cd8e9a4`](https://github.com/apache/spark/commit/cd8e9a4ddf704f2d01df870cc898af53e62a9d2f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65205/
Test PASSed.

srowen commented:
I'm OK with this as minor cleanup. It can't hurt.

srowen commented:
Merged to master
